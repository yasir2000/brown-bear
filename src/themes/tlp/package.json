{
  "author": "Enalean Team",
  "name": "@tuleap/tlp",
  "homepage": "https://tuleap.org",
  "license": "GPL-2.0-or-later",
  "description": "Tuleap CSS framework",
  "private": true,
  "version": "0.0.0",
  "module": "./src/js/index.ts",
  "dependencies": {
    "@tuleap/tlp-fetch": "file:../../scripts/lib/tlp-fetch",
    "@tuleap/tlp-popovers": "file:../../scripts/lib/tlp-popovers",
    "flatpickr": "^4.6.9",
    "jquery": "^3.6.0",
    "select2": "^4.0.13"
  },
  "devDependencies": {
    "@types/jquery": "^3.5.5"
  },
  "scripts": {
    "build": "echo 'I exist only to make pnpm build things that depend on me in the right order'",
    "test": "jest"
  }
}
