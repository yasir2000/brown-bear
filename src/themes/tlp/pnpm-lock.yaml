lockfileVersion: 5.3

specifiers:
  '@tuleap/tlp-fetch': file:../../scripts/lib/tlp-fetch
  '@tuleap/tlp-popovers': file:../../scripts/lib/tlp-popovers
  '@types/jquery': ^3.5.5
  flatpickr: ^4.6.9
  jquery: ^3.6.0
  select2: ^4.0.13

dependencies:
  '@tuleap/tlp-fetch': link:../../scripts/lib/tlp-fetch
  '@tuleap/tlp-popovers': link:../../scripts/lib/tlp-popovers
  flatpickr: 4.6.9
  jquery: 3.6.0
  select2: 4.0.13

devDependencies:
  '@types/jquery': 3.5.5

packages:

  /@types/jquery/3.5.5:
    resolution: {integrity: sha512-6RXU9Xzpc6vxNrS6FPPapN1SxSHgQ336WC6Jj/N8q30OiaBZ00l1GBgeP7usjVZPivSkGUfL1z/WW6TX989M+w==}
    dependencies:
      '@types/sizzle': 2.3.2
    dev: true

  /@types/sizzle/2.3.2:
    resolution: {integrity: sha512-7EJYyKTL7tFR8+gDbB6Wwz/arpGa0Mywk1TJbNzKzHtzbwVmY4HR9WqS5VV7dsBUKQmPNr192jHr/VpBluj/hg==}
    dev: true

  /flatpickr/4.6.9:
    resolution: {integrity: sha512-F0azNNi8foVWKSF+8X+ZJzz8r9sE1G4hl06RyceIaLvyltKvDl6vqk9Lm/6AUUCi5HWaIjiUbk7UpeE/fOXOpw==}
    dev: false

  /jquery/3.6.0:
    resolution: {integrity: sha512-JVzAR/AjBvVt2BmYhxRCSYysDsPcssdmTFnzyLEts9qNwmjmu4JTAMYubEfwVOSwpQ1I1sKKFcxhZCI2buerfw==}
    dev: false

  /select2/4.0.13:
    resolution: {integrity: sha512-1JeB87s6oN/TDxQQYCvS5EFoQyvV6eYMZZ0AeA4tdFDYWN3BAGZ8npr17UBFddU0lgAt3H0yjX3X6/ekOj1yjw==}
    dev: false
