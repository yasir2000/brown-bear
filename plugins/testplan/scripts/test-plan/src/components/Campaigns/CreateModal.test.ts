/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type Vue from "vue";
import { shallowMount } from "@vue/test-utils";
import { createTestPlanLocalVue } from "../../helpers/local-vue-for-test";
import * as tlp from "tlp";
import type { Modal } from "tlp";
import CreateModal from "./CreateModal.vue";
import { createStoreMock } from "../../../../../../../src/scripts/vue-components/store-wrapper-jest";
import type { RootState } from "../../store/type";
import type { CampaignState } from "../../store/campaign/type";
import * as tracker_report_retriever from "../../helpers/Campaigns/tracker-reports-retriever";
import type { BacklogItemState } from "../../store/backlog-item/type";

jest.mock("tlp", () => {
    return {
        __esModule: true,
        createModal: jest.fn(),
    };
});

describe("CreateModal", () => {
    let local_vue: typeof Vue;

    beforeEach(async () => {
        local_vue = await createTestPlanLocalVue();
    });

    it("Display the modal when mounted", async () => {
        const modal_show = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: modal_show,
            } as unknown as Modal;
        });

        const wrapper = shallowMount(CreateModal, {
            localVue: local_vue,
            mocks: {
                $store: createStoreMock({
                    state: {
                        milestone_title: "Milestone Title",
                        testdefinition_tracker_id: null,
                    } as RootState,
                }),
            },
        });
        // We need to wait for the loading state to be rendered and the tracker reports status to be resolved
        await wrapper.vm.$nextTick();

        expect(modal_show).toHaveBeenCalledTimes(1);
        expect(wrapper.element).toMatchSnapshot();
    });

    it("creates the campaign, hides the modal and refresh the backlog items", async () => {
        const modal_hide = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: jest.fn(),
                hide: modal_hide,
            } as unknown as Modal;
        });

        const $store = createStoreMock({
            state: {
                milestone_title: "Milestone Title",
                testdefinition_tracker_id: null,
                campaign: {} as CampaignState,
                backlog_item: {} as BacklogItemState,
            } as RootState,
        });

        const wrapper = shallowMount(CreateModal, {
            localVue: local_vue,
            mocks: {
                $store,
            },
        });

        // We need to wait for the loading state to be rendered and the tracker reports status to be resolved
        await wrapper.vm.$nextTick();

        wrapper.find("[data-test=new-campaign-label]").setValue("My new campaign");
        wrapper.vm.$data.test_selector = "milestone";

        await wrapper.trigger("submit");

        expect($store.dispatch).toHaveBeenCalledWith("campaign/createCampaign", {
            label: "My new campaign",
            initial_tests: {
                test_selector: "milestone",
            },
        });
        expect($store.dispatch).toHaveBeenCalledWith("backlog_item/loadBacklogItems");
        expect(modal_hide).toHaveBeenCalledTimes(1);
    });

    it("sets an error message when the reports of the test definition tracker cannot be retrieved", async () => {
        const modal_hide = jest.fn();
        jest.spyOn(tlp, "createModal").mockImplementation(() => {
            return {
                show: jest.fn(),
                hide: modal_hide,
            } as unknown as Modal;
        });
        jest.spyOn(tracker_report_retriever, "getTrackerReports").mockRejectedValueOnce(
            new Error("Something bad happened")
        );

        const $store = createStoreMock({
            state: {
                milestone_title: "Milestone Title",
                testdefinition_tracker_id: 12,
            } as RootState,
        });

        const wrapper = shallowMount(CreateModal, {
            localVue: local_vue,
            mocks: {
                $store,
            },
        });
        // We need to wait for the loading state to be rendered and then to fail to retrieve the tracker reports
        await wrapper.vm.$nextTick();
        await wrapper.vm.$nextTick();

        // We need to mock console.error otherwise vue-test-utils tell us something has gone wrong when mounting the
        // component because we re-throw the error which ends up failing the test
        const consoleErrorSpy = jest.spyOn(global.console, "error").mockImplementation();
        try {
            await wrapper.vm.$nextTick();
        } finally {
            expect(consoleErrorSpy).toHaveBeenCalled();
            consoleErrorSpy.mockRestore();
        }

        expect(wrapper.find("[data-test=new-campaign-error-message]").exists()).toBe(true);
    });
});
