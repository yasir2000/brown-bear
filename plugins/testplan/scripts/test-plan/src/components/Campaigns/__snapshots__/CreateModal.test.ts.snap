// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateModal Display the modal when mounted 1`] = `
<form
  aria-labelledby="test-plan-create-campaign-modal-title"
  class="tlp-modal"
  role="dialog"
>
  <div
    class="tlp-modal-header"
  >
    <h1
      class="tlp-modal-title"
      id="test-plan-create-campaign-modal-title"
    >
      <translate-stub
        tag="span"
      >
        Create new campaign
      </translate-stub>
    </h1>
     
    <button
      aria-label="Close"
      class="tlp-modal-close"
      data-dismiss="modal"
      type="button"
    >
      <i
        aria-hidden="true"
        class="fas fa-times tlp-modal-close-icon"
      />
    </button>
  </div>
   
  <!---->
   
  <div
    class="tlp-modal-body"
  >
    <div>
      <div
        class="tlp-form-element"
      >
        <label
          class="tlp-label"
          for="new-campaign-label"
        >
          <translate-stub
            tag="span"
          >
            Name
          </translate-stub>
           
          <i
            aria-hidden="true"
            class="fa fa-asterisk"
          />
        </label>
         
        <input
          class="tlp-input"
          data-test="new-campaign-label"
          id="new-campaign-label"
          required="required"
          type="text"
        />
      </div>
       
      <create-campaign-test-selector-stub
        testdefinition_tracker_reports=""
        value="[object Object]"
      />
    </div>
  </div>
   
  <div
    class="tlp-modal-footer"
  >
    <button
      class="tlp-button-primary tlp-button-outline tlp-modal-action"
      data-current-language="en_US"
      data-dismiss="modal"
      data-msgid="
            Cancel
        "
      type="button"
    >
      
            Cancel
        
    </button>
     
    <button
      class="tlp-button-primary tlp-modal-action"
      data-test="new-campaign-submit-button"
      type="submit"
    >
      <i
        aria-hidden="true"
        class="fa tlp-button-icon fa-save"
      />
       
      <translate-stub
        tag="span"
      >
        Create campaign
      </translate-stub>
    </button>
  </div>
</form>
`;
